test/abstract_syntax_tree.coffee

```diff
@@ -261,34 +710,246 @@ test "AST as expected for Elision node", ->
-  # TODO
+  testExpression 'Ghosts?',
+    type: 'Existence',
+    comparisonTarget: 'null'
+    expression: value: 'Ghosts'
+
+  # NOTE: Soaking is covered in `Call` and `Access` nodes.
 
 test "AST as expected for Parens node", ->
-  # TODO
+  testExpression '(hmmmmm)',
+    type: 'Parens',
+    body: type: 'Value'
+
+  testExpression '(a + b) / c',
+    type: 'Op'
+    operator: '/'
+    first:
+      type: 'Parens'
+      body:
+        type: 'Op'
+        operator: '+'
+
+  testExpression '(((1)))',
+    type: 'Parens',
+    body:
+      type: 'Value'
+      base:
+        type: 'Block'
+        expressions: [
+          type: 'Parens',
+          body:
+            type: 'Value'
+            base: value: '1'
+        ]
 
 test "AST as expected for StringWithInterpolations node", ->
-  # TODO
+  testExpression '"#{o}/"',
+    type: 'StringWithInterpolations'
+    quote: '"'
+    body:
+      type: 'Block'
+      expressions: [
+        originalValue: ''
+      ,
+        type: 'Interpolation'
+        expression:
+          type: 'Value'
+          base: value: 'o'
+      ,
+        originalValue: '/'
+      ]
 
 test "AST as expected for For node", ->
-  # TODO
+  testExpression 'for x, i in arr when x? then return',
+    type: 'For'
+    from: undefined
+    object: undefined
+    range: no
+    pattern: no
+    returns: no
+    guard: type: 'Existence'
+    source: type: 'IdentifierLiteral'
+    body: type: 'Return'
+
+  testExpression 'for k, v of obj then return',
+    type: 'For'
+    from: undefined
+    object: yes
+    range: no
+    pattern: no
+    returns: no
+    guard: undefined
+    source: type: 'IdentifierLiteral'
+
+  testExpression 'for x from iterable then',
+    type: 'For'
+    from: yes
+    object: undefined
+    body: type: 'Block'
+    source: type: 'IdentifierLiteral'
+
+  testExpression 'for i in [0...42] by step when not i % 2 then',
+    type: 'For'
+    from: undefined
+    object: undefined
+    range: yes
+    pattern: no
+    returns: no
+    body: type: 'Block'
+    source: type: 'Range'
+    guard: type: 'Op'
+    step: type: 'IdentifierLiteral'
+
+  testExpression 'a = (x for x in y)',
+    type: 'Assign'
+    value:
+      type: 'Parens'
+      body:
+        type: 'For'
+        returns: no
+        pattern: no
+
+  # TODO: Figgure out the purpose of `pattern` and `returns`.
 
 test "AST as expected for Switch node", ->
-  # TODO
+  testExpression 'switch x \n when a then a; when b, c then c else 42',
+    type: 'Switch'
+    subject:
+      type: 'IdentifierLiteral'
+      value: 'x'
+    cases: [
+      {type: 'IdentifierLiteral', value: 'a'}
+      {type: 'Value', base: value: 'a'}
+      {type: 'IdentifierLiteral', value: 'b'}
+      {type: 'IdentifierLiteral', value: 'c'}
+      {type: 'Value', base: value: 'c'}
+    ]
+    otherwise:
+      type: 'Value'
+      base: value: '42'
+      isDefaultValue: no
+
+  # TODO: File issue for compile error when using `then` or `;` where `\n` is rn.
```

Comment:
?

Reply:
From playing around for a bit it seemed impossible to enter the `Switch` body without an actual linebreak. From my understanding a semicolon or the `then` keyword should allow for it.